### PR TITLE
Feature/issue 3048 panic invalid update

### LINKED ISSUE #3048: Panic when attempting to make invalid update to a conflicting document
Steps to repro
1. Create a document
2. Update the document
3. Add a conflicting revision.  e.g. something like:
```
http --verbose PUT localhost:4985/default/conflictTest value=conflicted new_edits==false _revisions:='{"start":2, "ids":["conflicting_rev", "f6fa803508c40388de38c9f99729c835"]}'
```
4. Issue the same request again

Step 4 panics - it should just return a 409 conflict.

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #3048

### CODE REVIEW

**Comment by tleyden** [comment]
Passing integration tests for newly added tests:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/121/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/122/

**Comment by tleyden** [comment]
@adamcfraser this should be ready for review

File: rest/doc_api.go [hunk 1]
```diff
@@ -279,7 +279,12 @@ func (h *handler) handlePutDoc() error {
 		if err != nil {
 			return err
 		}
-		newRev = body["_rev"].(string)
+
+		newRevRaw, ok := body["_rev"]
+		if !ok {
+			return base.HTTPErrorf(http.StatusInternalServerError, "Expected revision id in body _rev field")
+		}
+		newRev = newRevRaw.(string)
```
Comment (adamcfraser) [inline rest/doc_api.go]: I think you could simplify this as 
```
newRev, ok := body["_rev"].(string)
if !ok {
   ...
}
```
https://play.golang.org/p/IMXQ_LdABB

File: rest/admin_api_test.go [hunk 2]
```diff
@@ -29,6 +29,60 @@ import (
...
+	defer rt.Close()
+
+	docId := "conflictTest"
+
+	// Create doc
+	response := rt.SendAdminRequest("PUT", fmt.Sprintf("/db/%s", docId), `{"value":"initial"}`)
+	response.DumpBody()
+
+	// Discover revision ID
+	// TODO: Is there any way to leverage db.document here to avoid the type assertions?
```
Comment (adamcfraser) [inline rest/admin_api_test.go]: We define a temporary struct for this kind of scenario in a lot of tests, but really this suggests to me that the schema for SG responses should be defined in our code somewhere, even if it's more economical to build it as a simple Body in places like handlePutDoc.

File: db/crud.go [hunk 3]
```diff
@@ -672,6 +672,7 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string)
 		}
 		if currentRevIndex == 0 {
 			base.LogTo("CRUD+", "PutExistingRev(%q): No new revisions to add", docid)
+			body["_rev"] = newRev
```
Comment (adamcfraser) [inline db/crud.go]: The scoping is very hard to follow here.  Below when we update body["_rev"] we're about to pass body back to updateDoc via the callback.  Here we're directly updating the body for use in the PutExistingRev scope.  I can see how it might work for this issue, but it's difficult to understand the intention (and why we don't set "_rev" in other cases).

I think it's ok to leave for now, though - maybe just with an additional comment for clarification.  We're planning to refactor out most of these usages of _rev and _id in the document body in 2.0 anyway, so we can revisit at that point.

**Review by adamcfraser — COMMENTED** [review]
Generally looks ok - a few comments

File: db/crud.go (hunk 3 above)
Comment (tleyden) [reply db/crud.go]: Adding a comment

File: rest/admin_api_test.go (hunk 2 above)
Comment (tleyden) [reply rest/admin_api_test.go]: Yeah, ok I'll leave as-is and update the TODO

File: rest/doc_api.go (hunk 1 above)
Comment (tleyden) [reply rest/doc_api.go]: This might be tricky.  I started out with something like that (not sure if it was exactly that), but it was creating a new variable rather than updating the existing `newRev` in scope, and leaving the outer `newRev` untouched.

To guard against that possibility, I reworked it to:

```
	        var newRev string
	        var ok bool
		newRev, ok = body["_rev"].(string)
		if !ok {
			return base.HTTPErrorf(http.StatusInternalServerError, "Expected revision id in body _rev field")
		}
```

I guess this is probably not idiomatic, but I've yet to come across a good way to remember the esoteric rules around shadowing and assignment in cases like these

File: rest/utilities_testing.go [hunk 4]
```diff
@@ -464,3 +464,19 @@ func (s *SlowResponseRecorder) Write(buf []byte) (int, error) {
 
 	return numBytesWritten, err
 }
+
+
+// Duplicated from DB package
```
[the remark this answers is not in the document]
Comment (tleyden) [reply rest/utilities_testing.go]: Oops, I got the cross-package dependency restrictions mixed up there.  Thanks for catching that.

**Comment by tleyden** [comment]
Changes pushed up
